tachyon/crypto/sumcheck/multilinear/data_structures.h

```diff
@@ -0,0 +1,173 @@
+      auto it = raw_pointers_lookup_table_.find(evaluation_ptr);
+      if (it != raw_pointers_lookup_table_.end()) {
+        indexed_evaluations.push_back(
+            raw_pointers_lookup_table_[evaluation_ptr]);
+      } else {
+        size_t curr_index = flattened_ml_evaluations_.size();
+        flattened_ml_evaluations_.push_back(evaluation);
+        raw_pointers_lookup_table_[evaluation_ptr] = curr_index;
+        indexed_evaluations.push_back(curr_index);
+      }
+    }
+    terms_.push_back(Term{coefficient, indexed_evaluations});
+  }
+
+  // Evaluates all terms together on a an evaluation point for each variable:
+  // $$\sum_{i=0}^{n}coefficient_i\cdot\prod_{j=0}^{m_i}evaluation_{ij}$$
+  F Evaluate(std::vector<F>& point) const {
+    CHECK_EQ(point.size(), num_variables_);
+    return std::accumulate(
+        terms_.begin(), terms_.end(), F::Zero(),
+        [this, &point](F& sum, const Term& term) {
+          return sum +=
+                 term.coefficient *
+                 std::accumulate(
+                     term.indexes.begin(), term.indexes.end(), F::One(),
+                     [this, &point](F& product, size_t index) {
+                       return product *=
+                              this->flattened_ml_evaluations_[index]->Evaluate(
+                                  point);
+                     });
+        });
+  }
+
+  // Extract the max number of evaluations and number of variables
+  PolynomialInfo Info() const { return {max_evaluations_, num_variables_}; }
+
+ private:
+  // max number of evaluations for each term
+  size_t max_evaluations_ = 0;
+  // number of variables of the polynomial
+  // Note(ashjeong): this is equivalent to `kMaxDegree` of multilinear
+  // extensions
+  size_t num_variables_ = 0;
+  // list of coefficients and their corresponding list of indexes in reference
+  // to evaluations in flattened_ml_evaluations_
+  std::vector<Term> terms_;
+  // holds shared pointers of multilinear evaluations
+  std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>
+      flattened_ml_evaluations_;
+  // holds all unique dense ML evaluations currently used + their corresponding
+  // index in "flattened_ml_evaluations_"
+  // not owned
+  absl::node_hash_map<const math::MultilinearExtension<Evaluations>*, size_t>
+      raw_pointers_lookup_table_;
+};
+
+// NOTE(ashjeong): univariate_evaluations.h used as reference
+template <typename H, typename F, typename Evaluations>
+H AbslHashValue(H h, const math::MultilinearExtension<Evaluations>* evals) {
+  size_t degree = 0;
+  for (const F& eval : evals->evaluations()) {
+    h = H::combine(std::move(h), eval);
+    ++degree;
+  }
+  F zero = F::Zero();
+  size_t max_degree = evals->Degree();
+  for (size_t i = degree; i <= max_degree; ++i) {
+    h = H::combine(std::move(h), zero);
+  }
+  return h;
+}
+
+}  // namespace crypto::sumcheck
+
+namespace base {
+
+using PolynomialInfo = crypto::sumcheck::PolynomialInfo;
```

Comment:
Please do not create type alias in header. This might cause unexpected symbol errors.